src/arch/xtensa/drivers/interrupt.c

```diff
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2019 Intel Corporation. All rights reserved.
+//
+// Author: Tomasz Lauda <tomasz.lauda@linux.intel.com>
+
+#include <sof/atomic.h>
+#include <sof/drivers/interrupt.h>
+#include <sof/lib/cpu.h>
+#include <xtensa/hal.h>
+#include <xtensa/xtruntime.h>
+#include <stdint.h>
+
+static struct arch_interrupt arch_irqs[PLATFORM_CORE_COUNT];
+
+void arch_interrupt_unregister(int irq)
+{
+	/* interrupt still enabled by some client */
+	if (atomic_read(&arch_irqs[cpu_get_id()].enabled_count[irq]))
```

Comment:
Yes, you are right. Would you like to implement the full functionality? I can revert this particular commit.

Reply:
Why revert? This hasn't been merged yet, right? So, no need to revert?

Reply:
I mean I can remove it from this patchset and let you do the full fix.

Reply:
@tlauda I see. In fact I think the correct fix should do a bit more than that. It seems to me, that repeatedly registering the same interrupt handler on the same CPU with the same argument is also a bug in the caller code. So, I think we should fix all callers and always return an error (-EBUSY) when arch_interrupt_register() is called for an already registered interrupt on the same CPU. Do you agree? Do we know which clients currently violate that?

Reply:
@lyakh This is exactly the thing I wanted to avoid. Currently the only client violating this is dw-dma. For some platforms the irq may be shared for all dma channels and all dmas of this type, for other it's not necessary true, but it doesn't mean we should fail. On cAVS platforms this is resolved with cascade interrupts. I don't see any reason, why similar approach couldn't be used for all the irqs. Allowing different clients to register for the same direct cpu irq is a nice feature to have.

Reply:
@tlauda it isn't only a nice feature to have, it's a common practice of interrupt sharing. But isn't it still broken with this your patch set? dw-dma registers interrupts with the argument equal to the channel pointer. So, if you register the same interrupt on the same CPU for different channels, the argument will be overwritten, so only the latest channel will be served? My understanding is, that XTOS doesn't support this kind of interrupt sharing. The proper fix for the problem seems to be to implement that kind of sharing. I understand, that this would be an overkill for a class of legacy systems for just one user - dw-dma. A different possibility *might* be to implement a driver for the internal IRQ multiplexer in the DW DMA controller, using the cascading interrupt controller API. Not sure this would work, though. I think the state of the dw-dma driver before 99e9f4370efbfbfb ("dw-dma: unify irq handling for all platforms") was correct - on most platforms, where all channels are using the same IRQ, it was registering only one handler and scanning the channels internally. Looks like that commit broke non-cAVS platforms.

Reply:
@lyakh It's not broken, because for non-cAVS platforms the passed channel is ignored in the handler. The easiest fix would be to add mask to the dma and just register/unregister based on that, but it isn't the perfect solution. It should be totally transparent to the DMA. What with the approach, where every interrupt is treated as cascaded? That would resolve all our issues.

Reply:
@tlauda I don't think the channel is currently ignored:
```
#if CONFIG_HW_LLI
#if CONFIG_DMA_AGGREGATED_IRQ
	/* skip if channel is not registered on this core */
	mask = BIT(chan - p->chan);
#else
	mask = ~0;
#endif

	status_src = dw_read(dma, DW_STATUS_BLOCK) & mask;
```
and it was my commit f0e5094c4a7cb60f728abf86b5e5d3b30bd52bf0 that changed that. Looks like it wasn't a good idea for BYT?

What do you mean by "tread as cascaded?" You don't mean really installing cascading IRQ controller descriptors for each DSP interrupt? Implementing all the DSP interrupts as an instance of a cascading controller probably won't work with the existing XTOS code either. Or do you mean implementing IRQ sharing by adding lists of clients to each DSP interrupt? I proposed that above, that would work, but I think that's a bit too much to just fix DW DMA on BYT.

Reply:
@lyakh Yep, probably an overkill. I'm in the middle of reworking DMA scheduling anyway, so the problem with multiple registrations should be resolved for free. In the meantime, I will remove this particular commit from this PR and just leave rest of them.